### PR TITLE
feat: [CDS-77806]: support moving inline services to remote

### PR DESCRIPTION
**Pull request by hingerabhinav**
## Describe your changes
support moving inline services to remote
provided a new API which takes in git params and service details. we create the file on remote and update the git details within harness db. 

Scenarios tested. 
1. moving service to default branch of the repo
2. moving service to non default branch of the repo 
3. error handling when scm errors occur

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55217)

### CODE REVIEW

**Comment by hingerabhinav** [comment]
trigger uts

**Review by adivishy1 — CHANGES_REQUESTED** [review]
> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55217)* status: 9 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java` line 1098 at r1](https://reviewable.io/reviews/harness/harness-core/55217#-NirVo6n4WphzNoL_eZr:-NirVo6n4WphzNoL_eZs:b-tjdfvr) ([raw file](https://github.com/harness/harness-core/blob/eeb71e130b65a7f440d3d367bcc52b68a74da21f/120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java#L1098)):*
> ```Java
>       })
>   @Hidden
>   public ResponseDTO<ServiceMoveConfigResponse>
> ```

are we adding it on the open api yaml as well?
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java` line 1111 at r1](https://reviewable.io/reviews/harness/harness-core/55217#-NirW92z6FhuoSff596-:-NirW92z6FhuoSff5960:baaqicl) ([raw file](https://github.com/harness/harness-core/blob/eeb71e130b65a7f440d3d367bcc52b68a74da21f/120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java#L1111)):*
> ```Java
>         Resource.of(NGResourceType.SERVICE, serviceIdentifier), SERVICE_UPDATE_PERMISSION);
>     ServiceMoveConfigResponse serviceMoveConfigResponse = serviceEntityService.moveServiceStoreTypeConfig(
>         accountIdentifier, orgIdentifier, projectIdentifier, serviceIdentifier, serviceRequestDTO);
> ```

ideally we should create a new DTO to pass the data from the resource layer to the service layer
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceRepositoryCustomImpl.java` line 434 at r1](https://reviewable.io/reviews/harness/harness-core/55217#-NirWfYnENeQBqDEqR0q:-NirWfYnENeQBqDEqR0r:b55z3hg) ([raw file](https://github.com/harness/harness-core/blob/eeb71e130b65a7f440d3d367bcc52b68a74da21f/127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceRepositoryCustomImpl.java#L434)):*
> ```Java
> 
>       // check whether GitX enabled for project
>       if (!cdGitXService.isNewGitXEnabled(
> ```

why dont do this check earlier in the api flow? we can fail fast

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/entity/ServiceMoveConfigRequestDTO.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  String baseBranch;
+
+  @Parameter(description = GitSyncApiConstants.MOVE_CONFIG_PARAM_MESSAGE)
+  @QueryParam(GitSyncApiConstants.MOVE_CONFIG_KEY)
+  MoveConfigOperationType moveConfigOperationType;
+
+  @Parameter(description = "Version Label")
+  @QueryParam(NGCommonEntityConstants.VERSION_LABEL_KEY)
+  @NotNull
+  String versionLabel;
```
Comment (tchaurasiya) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/entity/ServiceMoveConfigRequestDTO.java:66]: is it needed for service?

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java [hunk 2]
```diff
@@ -1517,4 +1520,38 @@ public RepoListResponseDTO getListOfRepos(String accountIdentifier, String orgId
 
     return RepoListResponseDTO.builder().repositories(uniqueRepos).build();
   }
+
+  @Override
+  public ServiceMoveConfigResponse moveServiceStoreTypeConfig(String accountIdentifier, String orgIdentifier,
+      String projectIdentifier, String serviceIdentifier, ServiceMoveConfigRequestDTO serviceMoveConfigRequestDTO) {
```
Comment (tchaurasiya) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java:1526]: can we add basic validation on request dto here (if any needed)

**Review by hingerabhinav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55217)* status: 11 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java` line 1098 at r1](https://reviewable.io/reviews/harness/harness-core/55217#-NirVo6n4WphzNoL_eZr:-Nistp5Q6PJg1fqaEWaN:b-i7ioyl) ([raw file](https://github.com/harness/harness-core/blob/eeb71e130b65a7f440d3d367bcc52b68a74da21f/120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java#L1098)):*
<details><summary><i>Previously, adivishy1 (Adithya Viswanathan) wrote…</i></summary><blockquote>

are we adding it on the open api yaml as well?
</blockquote></details>

not right now. there are a lot of APIs not present on openAPI. We can pick them up together. 
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java` line 1111 at r1](https://reviewable.io/reviews/harness/harness-core/55217#-NirW92z6FhuoSff596-:-NistvZx7Tfc1lyziYjP:bs7wumx) ([raw file](https://github.com/harness/harness-core/blob/eeb71e130b65a7f440d3d367bcc52b68a74da21f/120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java#L1111)):*
<details><summary><i>Previously, adivishy1 (Adithya Viswanathan) wrote…</i></summary><blockquote>

ideally we should create a new DTO to pass the data from the resource layer to the service layer
</blockquote></details>

Done. 
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/entity/ServiceMoveConfigRequestDTO.java` line 66 at r1](https://reviewable.io/reviews/harness/harness-core/55217#gh-1389051740:-NistzbL6ybC_1w3D3od:b-ybfw70) ([raw file](https://github.com/harness/harness-core/blob/eeb71e130b65a7f440d3d367bcc52b68a74da21f/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/entity/ServiceMoveConfigRequestDTO.java#L66)):*
<details><summary><i>Previously, tchaurasiya (Tathagat Chaurasiya) wrote…</i></summary><blockquote>

is it needed for service?
</blockquote></details>

Right. Its not. 
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java` line 1526 at r1](https://reviewable.io/reviews/harness/harness-core/55217#gh-1389054790:-Nisu9i4AmQt8b2XWR-u:b-dgeecv) ([raw file](https://github.com/harness/harness-core/blob/eeb71e130b65a7f440d3d367bcc52b68a74da21f/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java#L1526)):*
<details><summary><i>Previously, tchaurasiya (Tathagat Chaurasiya) wrote…</i></summary><blockquote>

can we add basic validation on request dto here (if any needed)
</blockquote></details>

Yes adding. 
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceRepositoryCustomImpl.java` line 434 at r1](https://reviewable.io/reviews/harness/harness-core/55217#-NirWfYnENeQBqDEqR0q:-Nisu8lc2kZC23lg6HWk:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/eeb71e130b65a7f440d3d367bcc52b68a74da21f/127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceRepositoryCustomImpl.java#L434)):*
<details><summary><i>Previously, adivishy1 (Adithya Viswanathan) wrote…</i></summary><blockquote>

why dont do this check earlier in the api flow? we can fail fast
</blockquote></details>

Done.

**Review by tchaurasiya — APPROVED** [review]

**Review by adivishy1 — APPROVED** [review]
> Reviewed 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55217)* status: 11 file(s) to review. 2. unresolved discussions.

**Comment by hingerabhinav** [comment]
trigger review

**Pull request merged**